code/modules/research/xenobiology/crossbreeding/warping.dm

```diff
@@ -0,0 +1,1227 @@
+	move_resist = INFINITY  //here to avoid the rune being moved since it only sets it's turf once when it's drawn. doesn't include admin fuckery.
+	anchored = TRUE
+	layer = MID_TURF_LAYER
+	resistance_flags = FIRE_PROOF
+	///is only used for bluespace crystal erasing as of now
+	var/storing_time = 5
+	///Nearly all runes needs to know which turf they are on
+	var/turf/rune_turf
+	///starting cooldown of the rune.
+	var/cooldown = 0
+	///duration of the cooldown for the rune only applies to certain runes
+	var/max_cooldown = 100
+
+/obj/item/slimecross/warping/Initialize()
+	. = ..()
+	desc = "It just won't stay in place. it has [warp_charge] charge left"
+
+
+
+///runes can also be deleted by bluespace crystals relatively fast as an alternative to cleaning them.
+/obj/effect/warped_rune/attackby(obj/item/used_item, mob/user)
+	. = ..()
+	if(!istype(used_item,/obj/item/stack/sheet/bluespace_crystal) && !istype(used_item,/obj/item/stack/ore/bluespace_crystal))
+		return
+
+	var/obj/item/stack/space_crystal = used_item
+	if(do_after(user, storing_time,target = src)) //the time it takes to nullify it depends on the rune too
+		to_chat(user, "<span class='notice'>You nullify the effects of the rune with the bluespace crystal!</span>")
+		qdel(src)
+		space_crystal.amount--
+		playsound(src, 'sound/effects/phasein.ogg', 20, TRUE)
+
+		if(space_crystal.amount <= 0)
+			qdel(space_crystal)
+
+
+/obj/effect/warped_rune/acid_act()
+	. = ..()
+	visible_message("<span class='warning'>[src] has been dissolved by the acid</span>")
+	playsound(src, 'sound/items/welder.ogg', 150, TRUE)
+	qdel(src)
+
+
+///nearly all runes use their turf in some way so we set rune_turf to their turf automatically, the rune also start on cooldown if it uses one.
+/obj/effect/warped_rune/Initialize()
+	. = ..()
+	rune_turf = get_turf(src)
+	RegisterSignal(rune_turf, COMSIG_COMPONENT_CLEAN_ACT, .proc/clean_rune)
+	cooldown = world.time + max_cooldown
+
+
+/obj/effect/warped_rune/proc/clean_rune()
+	qdel(src)
+
+
+///using the extract on the floor will "draw" the rune.
+/obj/item/slimecross/warping/afterattack(atom/target, mob/user, proximity)
+	if(!proximity)
+		return
+
+	if(isturf(target) && locate(/obj/effect/warped_rune) in target) //check if the target is a floor and if there's a rune on said floor
+		to_chat(user, "<span class='warning'>There is already a bluespace rune here!</span>")
+		return
+
+	if(istype(target, runepath))       //checks if the target is a rune and then if you can store it
+		if(warp_charge >= max_charge)
+			to_chat(user, "<span class='warning'>[src] is already full!</span>")
+			return
+
+		if(do_after(user, storing_time,target = target) && warp_charge < max_charge)
+			to_chat(user, "<span class='notice'>You store the rune in [src].</span>")
+			qdel(target)
+			warp_charge++
+			desc = "It just won't stay in place. it has [warp_charge] charge left"
+			return
+
+	if(!istype(target,/turf/open))
+		return
+
+	if(isspaceturf(target))
```

Comment:
Do redefined procs instead of long istype checks.
Things such as `warping_slime_act()` works.